superset/utils/core.py

```diff
@@ -1273,7 +1280,33 @@ def is_adhoc_metric(metric: Metric) -> bool:
 
 
 def get_metric_name(metric: Metric) -> str:
-    return metric["label"] if is_adhoc_metric(metric) else metric  # type: ignore
+    """
+    Extract label from metric
+
+    :param metric: object to extract label from
+    :return: String representation of metric
+    :raises ValueError: if metric object is invalid
+    """
+    if is_adhoc_metric(metric):
```

Comment:
Could we make `is_adhoc_metric` a [type guard](https://www.python.org/dev/peps/pep-0647/) so that we don't need to use `cast` and `# type: ignore` so much? It's available in [typing_extensions](https://github.com/python/typing/blob/master/typing_extensions/README.rst).

Reply:
Amazing, I didn't know these were forthcoming! 🚀 Super cool, I'll implement right away. Btw, the functionality surrounding the `Metric` type is is pretty bad shape (e.g. `QueryObject` assumes the presence of a metric type consisting of a dict with only the property `label` that doesn't have a type definition, some geospatial viz plugins in `viz.py` apparently rely on `get_metric_name` to return `None` and `int` values in some cases), so I decided to break up the refactor into multiple smaller PRs to keep them manageable. So I will follow-up with some additional cleanup avoid this becoming XXXL.

Reply:
FYI adding support for type guards ended up being a deep rabbit hole involving upgrading `mypy`, needing to change `pre-commit-hooks` confs and a few monitors worth of new linting errors. I'll merge this now and follow up with a PR that bumps `mypy` and adds support for type guards.